backends/fs/android/android-saf-fs.cpp

```diff
@@ -0,0 +1,698 @@
+Common::String AndroidSAFFilesystemNode::getName() const {
+	if (_safTree == nullptr || _safParent == nullptr) {
+		// _newName is for non-existent paths or root node pretty name
+		return _newName;
+	}
+
+	return lastPathComponent(_path, '/');
+}
+
+Common::String AndroidSAFFilesystemNode::getPath() const {
+	assert(_safTree != nullptr);
+
+	if (_safNode != nullptr) {
+		return _path;
+	}
+
+	// When no node, it means _path is the parent node
+	return _path + "/" + _newName;
+}
+
+AbstractFSNode *AndroidSAFFilesystemNode::getChild(const Common::String &n) const {
+	assert(_safTree != nullptr);
+	assert(_safNode != nullptr);
+
+	// Make sure the string contains no slashes
+	assert(!n.contains('/'));
+
+	JNIEnv *env = JNI::getEnv();
+
+	jstring name = env->NewStringUTF(n.c_str());
+
+	jobject child = env->CallObjectMethod(_safTree, _MID_getChild, _safNode, name);
+
+	env->DeleteLocalRef(name);
+
+	if (env->ExceptionCheck()) {
+		LOGE("SAFFSTree::getChild failed");
+
+		env->ExceptionDescribe();
+		env->ExceptionClear();
+
+		return nullptr;
+	}
+
+	if (child) {
+		AndroidSAFFilesystemNode *ret = new AndroidSAFFilesystemNode(_safTree, child);
+		env->DeleteLocalRef(child);
+		return ret;
+	}
+
+	return new AndroidSAFFilesystemNode(_safTree, _safNode, _path, n);
+}
+
+bool AndroidSAFFilesystemNode::getChildren(AbstractFSList &myList, ListMode mode,
+        bool hidden) const {
+	assert(_flags & DIRECTORY);
+
+	assert(_safTree != nullptr);
+	if (!_safNode) {
+		return false;
+	}
+
+	JNIEnv *env = JNI::getEnv();
+
+	jobjectArray array =
+	    (jobjectArray)env->CallObjectMethod(_safTree, _MID_getChildren, _safNode);
+
+	if (env->ExceptionCheck()) {
+		LOGE("SAFFSTree::getChildren failed");
+
+		env->ExceptionDescribe();
+		env->ExceptionClear();
+
+		return false;
+	}
+
+	myList.clear();
+
+	jsize size = env->GetArrayLength(array);
+	myList.reserve(size);
+
+	for (jsize i = 0; i < size; ++i) {
+		jobject node = env->GetObjectArrayElement(array, i);
+
+		myList.push_back(new AndroidSAFFilesystemNode(_safTree, node));
+
+		env->DeleteLocalRef(node);
+	}
+	env->DeleteLocalRef(array);
+
+	return true;
+}
+
+AbstractFSNode *AndroidSAFFilesystemNode::getParent() const {
+	assert(_safTree != nullptr);
+	assert(_safNode != nullptr);
+
+	if (_safParent) {
+		return new AndroidSAFFilesystemNode(_safTree, _safParent);
+	}
+
+	return AndroidFilesystemFactory::instance().makeRootFileNode();
+}
+
+Common::SeekableReadStream *AndroidSAFFilesystemNode::createReadStream() {
+	assert(_safTree != nullptr);
+
+	if (!_safNode) {
+		return nullptr;
+	}
+
+	JNIEnv *env = JNI::getEnv();
+
+	jint fd = env->CallIntMethod(_safTree, _MID_createReadStream, _safNode);
+
+	if (env->ExceptionCheck()) {
+		LOGE("SAFFSTree::createReadStream failed");
+
+		env->ExceptionDescribe();
+		env->ExceptionClear();
+
+		return nullptr;
+	}
+
+	if (fd == -1) {
+		return nullptr;
+	}
+
+	FILE *f = fdopen(fd, "r");
+	if (!f) {
+		close(fd);
+		return nullptr;
+	}
+
+	return new StdioStream(f);
```

Comment:
Would `PosixIoStream` work here as well?

Reply:
It seems to work.
I tested to detect a game and it worked.